asio/include/asio/fiber.hpp

```diff
@@ -0,0 +1,1553 @@
+  }
+
+  /// Reset the cancellation state of the fiber and insert in and out filters.
+  template <typename InFilter, typename OutFilter>
+  void reset_cancellation_state(ASIO_MOVE_ARG(InFilter) in_filter,
+                                ASIO_MOVE_ARG(OutFilter) out_filter)
+  {
+    handle_->cancel_state = cancellation_state(handle_->cancel_slot,
+                                              ASIO_MOVE_CAST(InFilter)(in_filter),
+                                              ASIO_MOVE_CAST(OutFilter)(out_filter));
+  }
+
+ public:
+
+  template<typename H, ASIO_COMPLETION_SIGNATURES_TPARAMS>
+  friend class async_result;
+
+  detail::fiber_handle *handle_;
+  // innter fiber
+  boost::context::fiber *fiber_;
+  executor_type executor_;
+
+};
+
+/// The fiber context type with the default executor.
+typedef basic_fiber_context<> fiber_context;
+
+namespace detail
+{
+
+template<typename Executor>
+struct initiate_fiber
+{
+
+  template<typename CompletionHandler, typename Func>
+  void operator()(CompletionHandler && completion_handler, Func && function, void(*)()) const
+  {
+    auto alloc = get_associated_executor(completion_handler, get_executor());
+
+    struct fiber_runner : fiber_handle
+    {
+      Executor executor;
+      Func func;
+      CompletionHandler handler;
+
+      fiber_runner(cancellation_slot slot, Executor executor, Func && func, CompletionHandler && handler)
+          : fiber_handle(slot), executor(std::move(executor)),
+            func(std::forward<Func>(func)), handler(std::forward<CompletionHandler>(handler)) {}
+
+      auto complete()
+      {
+        auto def = asio::experimental::deferred(
+                [this, lf = std::exchange(lifetime, nullptr)] () mutable
+                {
+                  return asio::experimental::deferred.template values();
+                });
+        if (this->did_suspend)
+          asio::dispatch(get_associated_executor(handler, this->executor), std::move(def))(std::move(handler));
+        else
+          asio::post(get_associated_executor(handler, this->executor), std::move(def))(std::move(handler));
+
+      }
+
+      auto work(boost::context::fiber && fb) noexcept -> boost::context::fiber
+      {
+        using fiber_context = basic_fiber_context<Executor>;
+        assert(lifetime);
+        func(fiber_context{*this, std::move(fb), executor});
+        complete();
+
+        return std::move(fb);
+      }
+
+      void start(std::unique_ptr<fiber_runner> p)
+      {
+        fiber = boost::context::fiber([this](boost::context::fiber && fb) {return work(std::move(fb));});
+        asio::dispatch(
+                this->executor,
+                [this, p = std::move(p)]() mutable
+                {
+                  lifetime = std::move(p);
+                  fiber = std::move(fiber).resume();
+                });
+
+      }
+    };
+
+    auto runner = std::make_unique<fiber_runner>(get_associated_cancellation_slot(completion_handler),
```

Comment:
`make_unique` is C++14?

Reply:
See 8190b17, I went down to C++14. `experimental::deferred` requires C++14 support.